File: worker/src/telegram.ts
/**
 * Telegram Bot API client
 */

import type { ParsedEvent, UserSubscription } from './types';

const TELEGRAM_API = 'https://api.telegram.org/bot';

const STATUS_EMOJI: Record<string, string> = {
  // Incident statuses
  investigating: '🔍',
  identified: '🔎',
  monitoring: '👀',
  resolved: '✅',
  postmortem: '📝',
  // Maintenance statuses
  scheduled: '📅',
  in_progress: '🔧',
  verifying: '🔎',
  completed: '✅',
};

const IMPACT_EMOJI: Record<string, string> = {
  none: '⚪',
  minor: '🟡',
  major: '🟠',
  critical: '🔴',
  maintenance: '🔵',
};

/**
 * Format an event into a Telegram message
 */
export function formatEventMessage(event: ParsedEvent, isUpdate = false): string {
  const statusEmoji = STATUS_EMOJI[event.status] || '🚨';
  const isMaintenance = event.eventType === 'maintenance';
  const impactEmoji = isMaintenance 
    ? IMPACT_EMOJI.maintenance 
    : (IMPACT_EMOJI[event.impact] || '⚪');

  let header: string;
  if (isUpdate) {
    header = '🔄 UPDATE';
  } else if (isMaintenance) {
    header = '🔧 SCHEDULED MAINTENANCE';
  } else {
    header = '🚨 NEW INCIDENT';
  }

  const lines: string[] = [
    header,
    '',
    `${statusEmoji} <b>${escapeHtml(event.name)}</b>`,
    `Status: ${event.status.replace('_', ' ').replace(/\b\w/g, c => c.toUpperCase())}`,
  ];

  // Show impact for incidents only
  if (!isMaintenance) {
    lines.push(`Impact: ${impactEmoji} ${event.impact.charAt(0).toUpperCase() + event.impact.slice(1)}`);
  }

  // Show scheduled time for maintenance
  if (isMaintenance && event.scheduled_for) {
    const start = event.scheduled_for.slice(0, 16).replace('T', ' ');
    const end = event.scheduled_until?.slice(11, 16) || '';
    lines.push(`⏰ Scheduled: ${start}${end ? ` → ${end}` : ''} UTC`);
  }

  // Add latest update body
  const latestUpdate = event.incident_updates?.[0];
  if (latestUpdate?.body) {
    let body = latestUpdate.body;
    if (body.length > 500) {
      body = body.slice(0, 497) + '...';
    }
    lines.push('', `📋 <i>${escapeHtml(body)}</i>`);
  }

  // Add affected components
  const componentNames = event.components?.slice(0, 5).map(c => c.name) || [];
  if (componentNames.length > 0) {
    lines.push('', `🎮 Affected: ${componentNames.join(', ')}`);
  }

  // Add link
  if (event.shortlink) {
    lines.push('', `🔗 ${event.shortlink}`);
  }

  return lines.join('\n');
}

/**
 * Escape HTML special characters
 */
function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

/**
 * Send a message via Telegram API
 */
export async function sendMessage(
  token: string,
  chatId: number,
  text: string,
  parseMode: 'HTML' | 'Markdown' = 'HTML'
): Promise<boolean> {
  const url = `${TELEGRAM_API}${token}/sendMessage`;
  
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        chat_id: chatId,
        text,
        parse_mode: parseMode,
        disable_web_page_preview: true,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      console.error(`Telegram API error for chat ${chatId}:`, error);
      return false;
    }

    return true;
  } catch (error) {
    console.error(`Failed to send message to ${chatId}:`, error);
    return false;
  }
}

/**
 * Send event notification to a user
 */
export async function sendEventNotification(
  token: string,
  chatId: number,
  event: ParsedEvent,
  isUpdate = false
): Promise<boolean> {
  const message = formatEventMessage(event, isUpdate);
  return sendMessage(token, chatId, message);
}

/**
 * Set webhook URL for the bot
 */
export async function setWebhook(token: string, url: string, secret?: string): Promise<boolean> {
  const apiUrl = `${TELEGRAM_API}${token}/setWebhook`;
  
  const body: Record<string, string> = { url };
  if (secret) {
    body.secret_token = secret;
  }

  try {
    const response = await fetch(apiUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });

    const result = await response.json() as { ok: boolean; description?: string };
    if (!result.ok) {
      console.error('Failed to set webhook:', result.description);
      return false;
    }
    return true;
  } catch (error) {
    console.error('Error setting webhook:', error);
    return false;
  }
}

/**
 * Delete webhook (for switching to polling mode)
 */
export async function deleteWebhook(token: string): Promise<boolean> {
  const url = `${TELEGRAM_API}${token}/deleteWebhook`;
  
  try {
    const response = await fetch(url, { method: 'POST' });
    const result = await response.json() as { ok: boolean };
    return result.ok;
  } catch (error) {
    console.error('Error deleting webhook:', error);
    return false;
  }
}

File: src/notifiers/telegram.py
"""Telegram notification backend."""

import os
import re

import requests

from ..epic_status import StatusEvent, EventType
from .base import Notifier


class TelegramNotifier(Notifier):
    """Send notifications via Telegram Bot API."""

    STATUS_EMOJI = {
        # Incident statuses
        "investigating": "🔍",
        "identified": "🔎",
        "monitoring": "👀",
        "resolved": "✅",
        "postmortem": "📝",
        # Maintenance statuses
        "scheduled": "📅",
        "in_progress": "🔧",
        "verifying": "🔎",
        "completed": "✅",
    }

    IMPACT_EMOJI = {
        "none": "⚪",
        "minor": "🟡",
        "major": "🟠",
        "critical": "🔴",
        "maintenance": "🔵",
    }

    def __init__(
        self,
        token: str | None = None,
        chat_id: str | None = None,
        timeout: int = 30,
    ):
        """
        Initialize Telegram notifier.
        
        Args:
            token: Bot token from @BotFather. Falls back to TELEGRAM_TOKEN env var.
            chat_id: Chat ID to send messages to. Falls back to TELEGRAM_CHAT_ID env var.
            timeout: Request timeout in seconds.
        """
        self.token = token or os.environ.get("TELEGRAM_TOKEN")
        self.chat_id = chat_id or os.environ.get("TELEGRAM_CHAT_ID")
        self.timeout = timeout

    @property
    def is_configured(self) -> bool:
        """Check if Telegram credentials are configured."""
        return bool(self.token and self.chat_id)

    def _format_message(self, event: StatusEvent, is_update: bool = False) -> str:
        """Format an event into a Telegram message."""
        status_emoji = self.STATUS_EMOJI.get(event.status, "🚨")
        
        # Use maintenance impact for maintenance events
        if event.is_maintenance:
            impact_emoji = self.IMPACT_EMOJI.get("maintenance", "🔵")
        else:
            impact_emoji = self.IMPACT_EMOJI.get(event.impact, "⚪")

        # Header based on event type and whether it's an update
        if is_update:
            header = "🔄 UPDATE"
        elif event.is_maintenance:
            header = "🔧 SCHEDULED MAINTENANCE"
        else:
            header = "🚨 NEW INCIDENT"

        lines = [
            header,
            "",
            f"{status_emoji} <b>{event.name}</b>",
            f"Status: {event.status.replace('_', ' ').title()}",
        ]
        
        # Show impact for incidents only
        if event.is_incident:
            lines.append(f"Impact: {impact_emoji} {event.impact.title()}")
        
        # Show scheduled time for maintenance
        if event.is_maintenance and event.scheduled_for:
            # Parse and format the time nicely
            scheduled = event.scheduled_for
            if event.scheduled_until:
                lines.append(f"⏰ Scheduled: {scheduled[:16].replace('T', ' ')} → {event.scheduled_until[11:16]} UTC")
            else:
                lines.append(f"⏰ Scheduled: {scheduled[:16].replace('T', ' ')} UTC")

        # Add latest update body if available
        if event.latest_update and event.latest_update.body:
            body = event.latest_update.body
            # Truncate long messages
            if len(body) > 500:
                body = body[:497] + "..."
            lines.append("")
            lines.append(f"📋 <i>{body}</i>")

        # Add affected components
        if event.component_names:
            components = ", ".join(event.component_names[:5])
            lines.append("")
            lines.append(f"🎮 Affected: {components}")

        # Add link
        if event.shortlink:
            lines.append("")
            lines.append(f"🔗 {event.shortlink}")

        return "\n".join(lines)

    def _send_message(self, message: str) -> bool:
        """Send a message via Telegram API."""
        if not self.is_configured:
            print("⚠️ Telegram not configured, printing message instead:")
            print("-" * 50)
            # Strip HTML tags for console output
            clean_message = re.sub(r'<[^>]+>', '', message)
            print(clean_message)
            print("-" * 50)
            return False

        url = f"https://api.telegram.org/bot{self.token}/sendMessage"
        payload = {
            "chat_id": self.chat_id,
            "text": message,
            "parse_mode": "HTML",
            "disable_web_page_preview": True,
        }

        try:
            resp = requests.post(url, json=payload, timeout=self.timeout)
            resp.raise_for_status()
            print("✅ Telegram message sent successfully")
            return True
        except (requests.RequestException, ConnectionError, TimeoutError) as e:
            print(f"❌ Failed to send Telegram message: {e}")
            return False

    def send_new_event(self, event: StatusEvent) -> bool:
        """Send notification for a new event."""
        message = self._format_message(event, is_update=False)
        return self._send_message(message)

    def send_event_update(self, event: StatusEvent) -> bool:
        """Send notification for an updated event."""
        message = self._format_message(event, is_update=True)
        return self._send_message(message)
